src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+    //TODO: Skill used
+
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+
+};
+
+process_id get_process_id( std::string id )
+{
+    return process_id( id );
+}
+
+generic_factory<process_data> processes_data( "process type", "id" );
+
+}
+
+void process_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "name", name );
+    //TODO: modify to allow lists of components
+    mandatory( jo, was_loaded, "components", components,  string_id_reader<itype_id> {});
+    mandatory( jo, was_loaded, "output", output );
+    mandatory( jo, was_loaded, "fuel_intake", fuel_intake );
+    mandatory( jo, was_loaded, "duration", duration );
+}
+
+void process_data::check() const
+{
+    const process_id proc_id( id.str() );
+
+    if( !proc_id.is_valid() ) {
+        debugmsg( "There is no process with the name %s", id.c_str() );
+    if( !components.is_valid() ) {
+        debugmsg( "Invalid components \"%s\" in \"%s\".", components.c_str(), id.c_str() );
+    }
+    if( !output.is_valid() ) {
+        debugmsg( "Invalid output \"%s\" in \"%s\".", output.c_str(), id.c_str() );
+    }
+    if( !fuel_intake.is_valid() ) {
+        debugmsg( "Invalid fuel intake \"%s\" in \"%s\".", fuel_intake.c_str(), id.c_str() );
+    }
+    if( !duration.is_valid() ) {
+        debugmsg( "Invalid duration \"%s\" in \"%s\".", duration.c_str(), id.c_str() );
+    }
+}
+
+void process_data::load( JsonObject &jo, const std::string &src )
+{
+    processes_data.load( jo, src );
+}
+
+void process_data::check()
+{
+    processes_data.check();
+}
+
+void process_data::reset()
+{
+    processes_data.reset();
+}
+
+void interact_with_processor( const tripoint &examp, player &p )
+{
+    //Get processor id at examp
+    processor_id pid = get_processor_id( examp );
+    //Do I need the debugmsg-es here or do they get checked in process_data::load?
+    if( !processors_data.is_valid( pid ) ) {
+        debugmsg( "Examined %s has action interact_with_processor, but has no processor data associated", g->m.furn( examp ).id().c_str() );
+        return;
+    }
+    //Get data associated with said processor id
+    const processor_data &processor = processors_data.obj( pid );
+    std::vector<auto> possible_processes;
+    for( auto &p : processor.processes() ) {
+        if( p.is_valid() ){
+            possible_processes.push_back( process_data.obj( process_id( p ) ) );
+        } else {
+            //Debugmsg probably not exactly right,
+            //I'm trying to return the processor name and the problematic process name too
+            debugmsg( "Examined %s has non-valid process data %s2 associated", processor.id().c_str(), get_process_id( p ).c_str() );
+            return;
+        }
+    }
+
+    //  Figure out which process to run, easier to work with than
+    //  trying to figure out desired process from items on tile,
+    //  Can become a mess with processor with multiple possible processes
```

Comment:
The problem is, selecting is only OK for starting processing.
If you can't figure out from items which process it was, just drop the multiple process option. It is not acceptable for processor to ask "so OK, which type of bricks were we making?" when it's done.